File: src/main/java/seremis/geninfusion/soul/Gene.java
package seremis.geninfusion.soul;

import seremis.geninfusion.api.soul.IAllele;
import seremis.geninfusion.api.soul.IChromosome;
import seremis.geninfusion.api.soul.IGene;
import seremis.geninfusion.api.soul.SoulHelper;
import seremis.geninfusion.soul.allele.*;

import java.util.Random;

public abstract class Gene implements IGene {

    protected Random rand = new Random();

    @Override
    public IChromosome mutate(IChromosome chromosome) {
        if(possibleAlleles() == AlleleBoolean.class) {
            AlleleBoolean allele1 = (AlleleBoolean) chromosome.getPrimary();
            AlleleBoolean allele2 = (AlleleBoolean) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                allele1.value = !allele1.value;
            } else {
                allele2.value = !allele2.value;
            }
        } else if(possibleAlleles() == AlleleInteger.class) {
            AlleleInteger allele1 = (AlleleInteger) chromosome.getPrimary();
            AlleleInteger allele2 = (AlleleInteger) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                allele1.value = (int) (allele1.value * (rand.nextFloat() * 2));
            } else {
                allele2.value = (int) (allele2.value * (rand.nextFloat() * 2));
            }
        } else if(possibleAlleles() == AlleleFloat.class) {
            AlleleFloat allele1 = (AlleleFloat) chromosome.getPrimary();
            AlleleFloat allele2 = (AlleleFloat) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                allele1.value = allele1.value * (rand.nextFloat() * 2);
            } else {
                allele2.value = allele2.value * (rand.nextFloat() * 2);
            }
        } else if(possibleAlleles() == AlleleDouble.class) {
            AlleleDouble allele1 = (AlleleDouble) chromosome.getPrimary();
            AlleleDouble allele2 = (AlleleDouble) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                allele1.value = allele1.value * (rand.nextFloat() * 2);
            } else {
                allele2.value = allele2.value * (rand.nextFloat() * 2);
            }
        } else if(possibleAlleles() == AlleleBooleanArray.class) {
            AlleleBooleanArray allele1 = (AlleleBooleanArray) chromosome.getPrimary();
            AlleleBooleanArray allele2 = (AlleleBooleanArray) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                int index = rand.nextInt(allele1.value.length);
                allele1.value[index] = !allele1.value[index];
            } else {
                int index = rand.nextInt(allele2.value.length);
                allele2.value[index] = !allele2.value[index];
            }
        } else if(possibleAlleles() == AlleleIntArray.class) {
            AlleleIntArray allele1 = (AlleleIntArray) chromosome.getPrimary();
            AlleleIntArray allele2 = (AlleleIntArray) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                int index = rand.nextInt(allele1.value.length);
                allele1.value[index] = (int) (allele1.value[index] * (rand.nextFloat() * 2));
            } else {
                int index = rand.nextInt(allele2.value.length);
                allele2.value[index] = (int) (allele2.value[index] * (rand.nextFloat() * 2));
            }
        } else if(possibleAlleles() == AlleleFloatArray.class) {
            AlleleFloatArray allele1 = (AlleleFloatArray) chromosome.getPrimary();
            AlleleFloatArray allele2 = (AlleleFloatArray) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                int index = rand.nextInt(allele1.value.length);
                allele1.value[index] = allele1.value[index] * (rand.nextFloat() * 2);
            } else {
                int index = rand.nextInt(allele2.value.length);
                allele2.value[index] = allele2.value[index] * (rand.nextFloat() * 2);
            }
        } else if(possibleAlleles() == AlleleDoubleArray.class) {
            AlleleDoubleArray allele1 = (AlleleDoubleArray) chromosome.getPrimary();
            AlleleDoubleArray allele2 = (AlleleDoubleArray) chromosome.getSecondary();

            if(rand.nextBoolean()) {
                int index = rand.nextInt(allele1.value.length);
                allele1.value[index] = allele1.value[index] * (rand.nextFloat() * 2);
            } else {
                int index = rand.nextInt(allele2.value.length);
                allele2.value[index] = allele2.value[index] * (rand.nextFloat() * 2);
            }
        }

        return chromosome;
    }

    @Override
    public IChromosome inherit(IChromosome chromosome1, IChromosome chromosome2) {
        IAllele allele1 = rand.nextBoolean() ? chromosome1.getPrimary() : chromosome1.getSecondary();
        IAllele allele2 = rand.nextBoolean() ? chromosome2.getPrimary() : chromosome2.getSecondary();

        return SoulHelper.instanceHelper.getIChromosomeInstance(allele1, allele2);
    }

    @Override
    public IChromosome advancedInherit(IChromosome[] parent1, IChromosome[] parent2, IChromosome[] offspring) {
        int geneId = SoulHelper.geneRegistry.getGeneId(this);

        IAllele allele1 = rand.nextBoolean() ? parent1[geneId].getPrimary() : parent1[geneId].getSecondary();
        IAllele allele2 = rand.nextBoolean() ? parent2[geneId].getPrimary() : parent2[geneId].getSecondary();

        return SoulHelper.instanceHelper.getIChromosomeInstance(allele1, allele2);
    }
}

File: src/main/java/seremis/geninfusion/soul/Gene.scala
package seremis.geninfusion.soul

import seremis.geninfusion.api.soul._

import scala.util.Random

class Gene(alleleType: EnumAlleleType) extends IGene {

    val rand = new Random()
    var mutate = true

    override def getAlleleType: EnumAlleleType = alleleType

    override def inherit(chromosome1: IChromosome, chromosome2: IChromosome): IChromosome = {
        val allele1 = if(rand.nextBoolean()) chromosome1.getPrimary else chromosome1.getSecondary
        val allele2 = if(rand.nextBoolean()) chromosome2.getPrimary else chromosome2.getSecondary

        SoulHelper.instanceHelper.getIChromosomeInstance(allele1, allele2)
    }

    override def advancedInherit(parent1: Array[IChromosome], parent2: Array[IChromosome], offspring: Array[IChromosome]): IChromosome = {
        val geneId = SoulHelper.geneRegistry.getGeneId(this)

        val allele1 = if(rand.nextBoolean()) parent1(geneId).getPrimary else parent1(geneId).getSecondary
        val allele2 = if(rand.nextBoolean()) parent2(geneId).getPrimary else parent2(geneId).getSecondary

        SoulHelper.instanceHelper.getIChromosomeInstance(allele1, allele2)
    }

    override def mutate(chromosome: IChromosome): IChromosome = {
        if(mutate) {
            var allele1Data = chromosome.getPrimary.getAlleleData
            var allele2Data = chromosome.getSecondary.getAlleleData

            alleleType match {
                case EnumAlleleType.BOOLEAN =>
                    if(rand.nextBoolean())
                        allele1Data != allele1Data.asInstanceOf[Boolean]
                    else
                        allele2Data != allele2Data.asInstanceOf[Boolean]
                case EnumAlleleType.BYTE =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Byte] * (rand.nextFloat() * 2)
                    else
                        allele2Data = allele2Data.asInstanceOf[Byte] * (rand.nextFloat() * 2)
                case EnumAlleleType.SHORT =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Short] * (rand.nextFloat() * 2)
                    else
                        allele2Data = allele2Data.asInstanceOf[Short] * (rand.nextFloat() * 2)
                case EnumAlleleType.INTEGER =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Int] * (rand.nextFloat() * 2)
                    else
                        allele2Data = allele2Data.asInstanceOf[Int] * (rand.nextFloat() * 2)
                case EnumAlleleType.FLOAT =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Float] * (rand.nextFloat() * 2)
                    else
                        allele2Data = allele2Data.asInstanceOf[Float] * (rand.nextFloat() * 2)
                case EnumAlleleType.DOUBLE =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Double] * (rand.nextFloat() * 2)
                    else
                        allele2Data = allele2Data.asInstanceOf[Double] * (rand.nextFloat() * 2)
                case EnumAlleleType.LONG =>
                    if(rand.nextBoolean())
                        allele1Data = allele1Data.asInstanceOf[Long] * (rand.nextFloat() * 2)
                    else
                        allele2Data = allele2Data.asInstanceOf[Long] * (rand.nextFloat() * 2)
                case EnumAlleleType.BOOLEAN_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Boolean]].length)
                        allele1Data.asInstanceOf[Array[Boolean]](index) != allele1Data.asInstanceOf[Array[Boolean]](index)
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Boolean]].length)
                        allele2Data.asInstanceOf[Array[Boolean]](index) != allele2Data.asInstanceOf[Array[Boolean]](index)
                    }
                case EnumAlleleType.BYTE_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Byte]].length)
                        allele1Data.asInstanceOf[Array[Byte]](index) = (allele1Data.asInstanceOf[Array[Byte]](index) * (rand.nextFloat() * 2)).toByte
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Byte]].length)
                        allele2Data.asInstanceOf[Array[Byte]](index) = (allele2Data.asInstanceOf[Array[Byte]](index) * (rand.nextFloat() * 2)).toByte
                    }
                case EnumAlleleType.SHORT_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Short]].length)
                        allele1Data.asInstanceOf[Array[Short]](index) = (allele1Data.asInstanceOf[Array[Short]](index) * (rand.nextFloat() * 2)).toShort
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Short]].length)
                        allele2Data.asInstanceOf[Array[Short]](index) = (allele2Data.asInstanceOf[Array[Short]](index) * (rand.nextFloat() * 2)).toShort
                    }
                case EnumAlleleType.INTEGER_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Int]].length)
                        allele1Data.asInstanceOf[Array[Int]](index) = (allele1Data.asInstanceOf[Array[Int]](index) * (rand.nextFloat() * 2)).toInt
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Int]].length)
                        allele2Data.asInstanceOf[Array[Int]](index) = (allele2Data.asInstanceOf[Array[Int]](index) * (rand.nextFloat() * 2)).toInt
                    }
                case EnumAlleleType.FLOAT_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Float]].length)
                        allele1Data.asInstanceOf[Array[Float]](index) = allele1Data.asInstanceOf[Array[Float]](index) * (rand.nextFloat() * 2)
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Float]].length)
                        allele2Data.asInstanceOf[Array[Float]](index) = allele2Data.asInstanceOf[Array[Float]](index) * (rand.nextFloat() * 2)
                    }
                case EnumAlleleType.DOUBLE_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Double]].length)
                        allele1Data.asInstanceOf[Array[Double]](index) = allele1Data.asInstanceOf[Array[Double]](index) * (rand.nextFloat() * 2)
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Double]].length)
                        allele2Data.asInstanceOf[Array[Double]](index) = allele2Data.asInstanceOf[Array[Double]](index) * (rand.nextFloat() * 2)
                    }
                case EnumAlleleType.LONG_ARRAY =>
                    if(rand.nextBoolean()) {
                        val index = rand.nextInt(allele1Data.asInstanceOf[Array[Long]].length)
                        allele1Data.asInstanceOf[Array[Long]](index) = (allele1Data.asInstanceOf[Array[Long]](index) * (rand.nextFloat() * 2)).toLong
                    } else {
                        val index = rand.nextInt(allele2Data.asInstanceOf[Array[Long]].length)
                        allele2Data.asInstanceOf[Array[Long]](index) = (allele2Data.asInstanceOf[Array[Long]](index) * (rand.nextFloat() * 2)).toLong
                    }
                case _ =>

            }
        }
        chromosome
    }

    override def noMutations(): IGene = {
        mutate = false
        this
    }
}
